packages/realm-react/src/cachedCollection.ts

```diff
@@ -0,0 +1,165 @@
+ * references on any relevant change (update, insert, deletion) and return the same
+ * object reference if no changes have occurred since the last access.
+ *
+ * This makes the {@link Realm.Collection} behave in an immutable way, as React expects, so
+ * that a {@link Realm.Object} can be wrapped in {@link React.memo} to prevent unnecessary
+ * rendering (see `useCollection` hook).
+ *
+ * @param args {@link CachedCollectionArgs} object arguments
+ * @returns Proxy object wrapping the collection
+ */
+export function cachedCollection<T extends Realm.Object>({
+  collection,
+  updateCallback,
+  collectionCache = new Map(),
+  isFirst = true,
+}: {
+  collection: Realm.Collection<T>;
+  updateCallback: () => void;
+  collectionCache?: Map<string, T>;
+  isFirst?: boolean;
+}): { collection: Realm.Collection<T>; tearDown: () => void } {
+  const cachedCollectionHandler: ProxyHandler<Realm.Collection<T & Realm.Object>> = {
+    get: function (target, key) {
+      // Pass functions through
+      const value = Reflect.get(target, key);
+      if (typeof value === "function") {
+        if (key === "sorted" || key === "filtered") {
+          return (...args: unknown[]) => {
+            const col: Realm.Results<T & Realm.Object> = Reflect.apply(value, target, args);
+            const { collection: newCol } = cachedCollection({
+              collection: col,
+              updateCallback,
+              collectionCache,
+              isFirst: false,
+            });
+            return newCol;
+          };
+        }
+        return value.bind(target);
+      }
+
+      // If the key is not numeric, pass it through
+      if (typeof key === "symbol" || !numericRegEx.test(key)) {
+        return value;
+      }
+
+      // If the key is numeric, check if we have a cached object for this key
+      const index = Number(key);
+      const object = target[index];
+      const objectId = object._objectId();
+      const cacheKey = getCacheKey(objectId);
+
+      // If we do, return it...
+      if (collectionCache.get(cacheKey)) {
+        return collectionCache.get(cacheKey);
+      }
+
+      // If not then this index has either not been accessed before, or has been invalidated due
+      // to a modification. Fetch it from the collection and store it in the cache
+      collectionCache.set(cacheKey, object);
+
+      return object;
+    },
+  };
+
+  const cachedCollectionResult = new Proxy(collection, cachedCollectionHandler);
+
+  const listenerCallback: Realm.CollectionChangeCallback<(T & Realm.Object) | (unknown & Realm.Object)> = (
+    listenerCollection,
+    changes,
+  ) => {
+    if (changes.deletions.length > 0 || changes.insertions.length > 0 || changes.newModifications.length > 0) {
+      // TODO: There is currently no way to rebuild the cache key from the changes array for deleted object.
+      // Until it is possible, we clear the cache on deletions.
+
+      // Possible solutions:
```

Comment:
add this [issue](https://github.com/realm/realm-core/issues/5220) to the comment